server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -672,16 +672,15 @@ final ReaderContext createAndPutReaderContext(ShardSearchRequest request, IndexS
                             + MAX_OPEN_SCROLL_CONTEXT.getKey() + "] setting.");
                 }
             }
-            final long keepAlive = getKeepAlive(request);
             final ShardSearchContextId id = new ShardSearchContextId(sessionId, idGenerator.incrementAndGet());
             if (keepStatesInContext || request.scroll() != null) {
-                readerContext = new LegacyReaderContext(id, indexService, shard, reader, request, keepAlive);
+                readerContext = new LegacyReaderContext(id, indexService, shard, reader, request, keepAliveInMillis);
                 if (request.scroll() != null) {
                     readerContext.addOnClose(decreaseScrollContexts);
                     decreaseScrollContexts = null;
                 }
             } else {
-                readerContext = new ReaderContext(id, indexService, shard, reader, keepAlive, request.keepAlive() == null);
+                readerContext = new ReaderContext(id, indexService, shard, reader, keepAliveInMillis, true);
```

Comment:
Should it be:

```suggestion
                readerContext = new ReaderContext(id, indexService, shard, reader, keepAliveInMillis, request.readerId() == null);
```
?

Reply:
I think all search contexts that we created here should be a single-session; otherwise, we will leak them when the query phase returns no result. We should only create multiple-session contexts when we explicitly [open a point in time](https://github.com/elastic/elasticsearch/blob/29192e591cb62f1b6ace13073fda23b448f69be3/server/src/main/java/org/elasticsearch/search/SearchService.java#L725). This is not an issue in the previous versions as we never create a search context when the readerId is specified.

Reply:
Right, thanks for clarifying.